src/config/test.rs

```diff
@@ -0,0 +1,184 @@
+        super::find_config_file(log, path)
+            .map_err(From::from)
+            .and_then(|s| Self::from_yaml(&s))
+    }
+
+    /// Attempts to deserialize [`Self`] from a YAML document. A valid source is
+    /// either a combination of [`Config`] document followed by [`TestConfig`]
+    /// document separated by a `---` (YAML document separator), or a
+    /// `TestConfig` document containing a `config` key that points to a valid
+    /// `Config` file.
+    pub fn from_yaml(src: &str) -> Result<Self, TestSuiteDecodeError> {
+        Ok(
+            if let Ok(options) = serde_yaml::from_str::<TestConfig>(src) {
+                let path = options
+                    .config
+                    .as_deref()
+                    .ok_or(TestSuiteDecodeError::MissingConfigInTestOptions)?;
+                let config = serde_yaml::from_reader(std::fs::File::open(path)?)?;
+                Self { config, options }
+            } else {
+                let mut de = serde_yaml::Deserializer::from_str(src);
+                let config =
+                    Config::deserialize(de.next().ok_or(TestSuiteDecodeError::MissingConfig)?)?;
+                let options = TestConfig::deserialize(
+                    de.next().ok_or(TestSuiteDecodeError::MissingTestOptions)?,
+                )?;
+                Self { config, options }
+            },
+        )
+    }
+}
+
+#[derive(Deserialize)]
+pub struct TestCase {
+    /// The data to be given to Quilkin.
+    pub input: Data,
+    /// What we expect Quilkin to send to the game server.
+    pub output: Data,
+    /// Metrics we expect to be set.
+    pub metrics: Option<std::collections::HashMap<String, MetricComparison>>,
+}
+
+#[derive(Deserialize)]
+pub struct MetricComparison {
+    pub name: String,
+    #[serde(deserialize_with = "metric_type_from_str")]
+    pub r#type: MetricType,
+    pub r#value: f64,
+}
+
+fn metric_type_from_str<'de, D>(deserializer: D) -> Result<MetricType, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    const METRIC_TYPES: &[(&str, MetricType)] = &[
+        ("counter", MetricType::COUNTER),
+        ("gauge", MetricType::GAUGE),
+        ("histogram", MetricType::HISTOGRAM),
+        ("summary", MetricType::SUMMARY),
+        ("untyped", MetricType::UNTYPED),
+    ];
+
+    let input = <&str>::deserialize(deserializer)?.to_lowercase();
+
+    METRIC_TYPES
+        .iter()
+        .find(|(key, _)| *key == input)
+        .map(|(_, value)| *value)
+        .ok_or_else(|| {
+            Error::custom(format!(
+                "Invalid Prometheus metric type. Expected: {}",
+                itertools::Itertools::intersperse(
+                    METRIC_TYPES.iter().map(|(key, _)| format!("`{}`", key)),
+                    String::from(", ")
+                )
+                .collect::<String>(),
+            ))
+        })
+}
+
+pub enum Data {
+    String(String),
+    Binary(Vec<u8>),
+    Base64(Vec<u8>),
+}
```

Comment:
what's this used for?

Reply:
This allows you accept strings, lists of numbers, or base64 as for the `input` and `output` properties, so you can write any of the following as valid data input for your tests.

```yaml
# Strings
input: Hello
# Bytes
input: [86, 45, 23, 23, 67]
# Base64
input:
   type: base64
   value: MWfu==

Reply:
Ah that makes sense!